File: jar/src/main/java/jline/api/mam/Mmap_rand.java
/**
 * @file Marked Markovian Arrival Process random generation
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

import java.util.Random;

public final class Mmap_rand {
    private Mmap_rand() {}

    /**
     * Generates a random MMAP with a given order and number of classes.
     *
     * @param order   the number of phases (order) in the MAP
     * @param classes the number of different classes (types) of arrivals in the MMAP
     * @return a MatrixCell representing the MMAP
     */
    public static MatrixCell mmap_rand(int order, int classes) {
        MatrixCell MMAP = new MatrixCell();
        for (int c = 0; c < 2 + classes; c++) {
            MMAP.set(c, new Matrix(order, order));
        }
        MatrixCell MAP = Map_rand.map_rand(order);

        MMAP.set(0, MAP.get(0));
        MMAP.set(1, MAP.get(1));

        Random rand = new Random();
        for (int i = 0; i < order; i++) {
            double[] p = new double[classes];
            double sum = 0.0;

            for (int c = 0; c < classes; c++) {
                p[c] = rand.nextDouble();
                sum += p[c];
            }

            for (int c = 0; c < classes; c++) {
                p[c] /= sum;
            }

            for (int j = 0; j < order; j++) {
                for (int c = 0; c < classes; c++) {
                    MMAP.set(2 + c, MMAP.get(1).copy());
                    MMAP.get(2 + c).scaleEq(p[c]);
                }
            }
        }

        return MMAP;
    }
}

File: jar/src/main/kotlin/jline/api/mam/Mmap_rand.kt
/**
 * @file Marked Markovian Arrival Process random generation
 * 
 * Generates random MMAP representations for testing and simulation purposes.
 * Essential for statistical validation and stochastic analysis of multiclass systems.
 * 
 * @since LINE 3.0
 */
package jline.api.mam

import jline.util.matrix.Matrix
import jline.util.matrix.MatrixCell
import java.util.*

/**
 * Generates a random MMAP (Marked Markovian Arrival Process) with a given order and number of classes.
 *
 * @param order   the number of phases (order) in the MAP
 * @param classes the number of different classes (types) of arrivals in the MMAP
 * @return a MatrixCell representing the MMAP, containing the D0, D1 matrices, and class-specific matrices
 */
fun mmap_rand(order: Int, classes: Int): MatrixCell {
    val MMAP = MatrixCell()
    for (c in 0..<2 + classes) {
        MMAP[c] = Matrix(order, order)
    }
    val MAP = map_rand(order)

    MMAP[0] = MAP[0]
    MMAP[1] = MAP[1]

    val rand = Random()
    for (i in 0..<order) {
        val p = DoubleArray(classes)
        var sum = 0.0

        for (c in 0..<classes) {
            p[c] = rand.nextDouble()
            sum += p[c]
        }

        for (c in 0..<classes) {
            p[c] /= sum
        }

        for (j in 0..<order) {
            for (c in 0..<classes) {
                MMAP[2 + c] = MMAP[1].copy()
                MMAP[2 + c].scaleEq(p[c])
            }
        }
    }

    return MMAP
}
/**
 * MMAP rand algorithms
 */
@Suppress("unused")
class MmapRandAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}